### PR TITLE
Add end of run restart functionality to MOM6

### LINKED ISSUE #118: ensure restart-at-end functionality is working in CMEPS
A configuration variable ``write_restart_at_endofrun`` is available in CMEPS. We need to test whether it is working and if not, find a solution. This may also involve the MOM6 and CICE6 Caps, which trigger restart-writing off the same configuration variables  used by CMEPS (``restart_n`` and ``restart_option``). Those components also need to have the same capability to write a restart at the end of the run.

### LINKED ISSUE #2236: Ensure restart-at-end functionality is working in CMEPS
## Description
This UFS WM issue is related to CMEPS issue https://github.com/NOAA-EMC/CMEPS/issues/118.
It will allow UFS WM to write out CMEPS restart files at the end of whole integration.

### PR DESCRIPTION
**Pull request by dpsarmie**
This PR allows the user to create restart files at the end of a run in MOM using the **write_restart_at_endofrun** configuration option in CMEPS. This configuration option will control end of run restarts for MOM6, CICE, and CMEPS. This PR closes NOAA-EMC/CMEPS#118 and closes ufs-community/ufs-weather-model#2236. A similar PR was made in the CICE repository ([CICE PR#77](https://github.com/NOAA-EMC/CICE/pull/77)) that will completely resolve these issues.

The code was tested on Hera using the regression test **datm_cdeps_control_gefs**. This was tested using different combinations of restart setting to try and ensure expected functionality. 
If the setting is not set or set to _False_, a restart file will not be created at the end of the run. Setting the option to _true_ will create the file.
The end of file restarts for CMEPS, CICE, and MOM will all be controlled with this single configuration option.

### CODE REVIEW

**Comment by jiandewang** [comment]
@DeniseWorthen can you take a preliminary code review before I reach out NCAR side ?

File: config_src/drivers/nuopc_cap/mom_cap.F90 [hunk 1]
```diff
@@ -648,6 +649,13 @@ subroutine InitializeAdvertise(gcomp, importState, exportState, clock, rc)
 
   endif
 
+  ! Read end of run restart config option
```
Comment (DeniseWorthen) [inline config_src/drivers/nuopc_cap/mom_cap.F90]: I think the retrieval of the config parameter is more naturally located in the InitializeP0 routine,  where the other config files are read.

File: config_src/drivers/nuopc_cap/mom_cap.F90 [hunk 2]
```diff
@@ -1807,7 +1816,18 @@ subroutine ModelAdvance(gcomp, rc)
...
     end if
 
-    if (write_restart .or. write_restartfh) then
+    write_restart_eor = .false.
+    if (restart_eor) then
+      if (ESMF_AlarmIsRinging(stop_alarm, rc=rc)) then
+         if (ChkErr(rc,__LINE__,u_FILE_u)) return
+         write_restart_eor = .true.
+         ! turn off the alarm
+         call ESMF_AlarmRingerOff(restart_alarm, rc=rc )
```
Comment (DeniseWorthen) [inline config_src/drivers/nuopc_cap/mom_cap.F90]: This needs to be turning off the stop_alarm, not the restart_alarm.

Comment (dpsarmie) [reply config_src/drivers/nuopc_cap/mom_cap.F90]: @DeniseWorthen Ok, code has been changed and tested.

**Comment by jiandewang** [comment]
@dpsarmie I made a try with latest UWM but replaed with your MOM6 branch. Somehow I am not getting what I expected. I used cpld_control_c48_intel as a template and changed restart_n from 12 to 18. The run length is 24hr and the IC is 20210306. With write_restart_at_endofrun=T it shall give us 20210400 and 20240406 ocean restart files. But I only see the frist one being written out. My run dir is on HERA: /scratch1/NCEPDEV/stmp2/Jiande.Wang/FV3_RT/rt_272814/cpld_control_c48_inte-T

can you take a look ? is there anything I missed or is there a specific CMEPS branch that I shall use ?
my UWM: /scratch1/NCEPDEV/climate/Jiande.Wang/working/scratch/MOM6-eor/ufs-weather-model, note here I replaced MOM6 with your branch

**Comment by DeniseWorthen** [comment]
@jiandewang I believe the test you want to do is to leave restart_n=12 but make the run length 27. You should get restarts at hour 12,24 and 27. Without this 'end-of-run' setting, you would only get restarts at 12,24, even though you run all the way to 27.

**Comment by jiandewang** [comment]
@DeniseWorthen in order to to extend to 27hr run, do I need to change stop_n to 27 (alone with model_configure nhours_fcst=27) ?

**Comment by DeniseWorthen** [comment]
@jiandewang stop_n should be set == to fhmax

**Comment by jiandewang** [comment]
@just made a quick test, see /scratch1/NCEPDEV/stmp2/Jiande.Wang/FV3_RT/rt_272814/cpld_control_c48_inte-TDW
restart_n = 12
stop_n = 30
nhours_fcst:             30

still don't see final restart file
/scratch1/NCEPDEV/stmp2/Jiande.Wang/FV3_RT/rt_272814/cpld_control_c48_inte-TDW[215]ls -l RESTART/*MOM*
-rw-r--r-- 1 Jiande.Wang stmp 6926138 May  8 13:52 RESTART/20210322.180000.MOM.res.nc
-rw-r--r-- 1 Jiande.Wang stmp 6926138 May  8 13:56 RESTART/20210323.060000.MOM.res.nc

**Comment by DeniseWorthen** [comment]
@dpsarmie Could you please check Jiande's run and see if you can spot the issue?

**Comment by dpsarmie** [comment]
@jiandewang _cpld_control_c48_ uses **parm/ufs.configure.s2s_esmf.IM** but that file does not have the configuration option active.

Go ahead and add **write_restart_at_endofrun = .true.** and rerun the test. That should (hopefully) solve the problem.

**Comment by dpsarmie** [comment]
@DeniseWorthen We can talk about whether or not we should add that option to the other ufs.configure files. Currently, only the HAFS and DATM ufs.configure files have the option in the configuration files.

**Comment by jiandewang** [comment]
@dpsarmie but if you see /scratch1/NCEPDEV/stmp2/Jiande.Wang/FV3_RT/rt_272814/cpld_control_c48_inte-TDW/ufs.configure, I have that in line 124

**Comment by dpsarmie** [comment]
> @dpsarmie but if you see /scratch1/NCEPDEV/stmp2/Jiande.Wang/FV3_RT/rt_272814/cpld_control_c48_inte-TDW/ufs.configure, I have that in line 124

Ok I see that. I'll try to run the c48 case and see if I can replicate the issue.

**Comment by DeniseWorthen** [comment]
@dpsarmie When we add this to UWM, we will want to add a configure variable to all the ufs.config files, but in the RT system, this could be false by default. The G-W will then be able to set it true if they need to.

**Comment by dpsarmie** [comment]
> @dpsarmie When we add this to UWM, we will want to add a configure variable to all the ufs.config files, but in the RT system, this could be false by default. The G-W will then be able to set it true if they need to.

Sounds good.

----

@jiandewang , I'm seeing that this is False in your mediator.log:
(med_phases_restart_alarm_init) write_restart_at_endofrun :       F (Line 537)

Haven't found what could be causing the issue causing the mediator flag to be incorrectly set but I'll keep looking tomorrow.

**Comment by DeniseWorthen** [comment]
I think the issue might be that Jiande used "true" vs ".true."

**Comment by dpsarmie** [comment]
I have a "true" case queued up on Hera right now. I figured that the parser would handle either case correctly but I'll wait an see what this test shows.

**Comment by jiandewang** [comment]
actually I tried .true. without success

**Comment by DeniseWorthen** [comment]
In CMEPS, it expects ".true."

https://github.com/NOAA-EMC/CMEPS/blob/4e19850cb083bc474b7cde5dc2f8506ec74cc442/mediator/med_phases_restart_mod.F90#L101-L106

**Comment by dpsarmie** [comment]
> actually I tried .true. without success

Ok. Denise is right though, I did a test run with "true" and it wasn't parsed correctly. I'll keep looking through your logs and see if there's any other issues.

**Comment by DeniseWorthen** [comment]
yes, just looking through some of the other logicals and it doesn't seem to matter. weird.

**Comment by jiandewang** [comment]
just re-submitted my test case with .true.
this time mediator.log shows 
(med_phases_restart_alarm_init) write_restart_at_endofrun :       T

let's wait for couple of minutes to see what's going on

**Comment by jiandewang** [comment]
using .true. give me what I am expecting.
now let me try run length=24, restart_n=18 to see what will happen

**Comment by DeniseWorthen** [comment]
@jiandewang Please check that you also get mediator cpl.r files at the same times that you get MOM6 restarts. In the end, when also including the CICE changes, we need all three components to have the capability to write at restart_n and at the end.

**Comment by jiandewang** [comment]
@DeniseWorthen yes we got that file
/scratch1/NCEPDEV/stmp2/Jiande.Wang/FV3_RT/rt_272814/cpld_control_c48_inte-TDW1/RESTART[119]ls -l *MOM*
-rw-r--r-- 1 Jiande.Wang stmp 6926138 May  9 10:03 20210322.180000.MOM.res.nc
-rw-r--r-- 1 Jiande.Wang stmp 6926138 May  9 10:07 20210323.060000.MOM.res.nc
-rw-r--r-- 1 Jiande.Wang stmp 6926138 May  9 10:10 20210323.120000.MOM.res.nc
/scratch1/NCEPDEV/stmp2/Jiande.Wang/FV3_RT/rt_272814/cpld_control_c48_inte-TDW1/RESTART[120]ls -l ufs.cpld.cpl*
-rw-r--r-- 1 Jiande.Wang stmp 8086484 May  9 10:03 ufs.cpld.cpl.r.2021-03-22-64800.nc
-rw-r--r-- 1 Jiande.Wang stmp 8086484 May  9 10:08 ufs.cpld.cpl.r.2021-03-23-21600.nc
-rw-r--r-- 1 Jiande.Wang stmp 8086484 May  9 10:10 ufs.cpld.cpl.r.2021-03-23-43200.nc

for my test case, run length=30, restart_n=12

**Comment by jiandewang** [comment]
my second try: run length=60, restart_n=18, works as expected:
/scratch1/NCEPDEV/stmp2/Jiande.Wang/FV3_RT/rt_272814/cpld_control_c48_inte-TDW3[185]ll RESTART/*MOM*
-rw-r--r-- 1 Jiande.Wang stmp 6926138 May  9 13:58 RESTART/20210323.000000.MOM.res.nc
-rw-r--r-- 1 Jiande.Wang stmp 6926138 May  9 14:05 RESTART/20210323.180000.MOM.res.nc
-rw-r--r-- 1 Jiande.Wang stmp 6926138 May  9 14:12 RESTART/20210324.120000.MOM.res.nc
-rw-r--r-- 1 Jiande.Wang stmp 6926138 May  9 14:14 RESTART/20210324.180000.MOM.res.nc

/scratch1/NCEPDEV/stmp2/Jiande.Wang/FV3_RT/rt_272814/cpld_control_c48_inte-TDW3[186]ll RESTART/ufs*
-rw-r--r-- 1 Jiande.Wang stmp 8086484 May  9 13:59 RESTART/ufs.cpld.cpl.r.2021-03-23-00000.nc
-rw-r--r-- 1 Jiande.Wang stmp 8086484 May  9 14:05 RESTART/ufs.cpld.cpl.r.2021-03-23-64800.nc
-rw-r--r-- 1 Jiande.Wang stmp 8086484 May  9 14:12 RESTART/ufs.cpld.cpl.r.2021-03-24-43200.nc
-rw-r--r-- 1 Jiande.Wang stmp 8086484 May  9 14:14 RESTART/ufs.cpld.cpl.r.2021-03-24-64800.nc

I am going to ask NACR to test on their side to make sure it won't break their system

**Comment by jiandewang** [comment]
@dpsarmie MOM6 dev/emc just had one updating. Can you sync your branch ? I will reach out NCAR after you sync your branch. Thanks

**Comment by dpsarmie** [comment]
> @dpsarmie MOM6 dev/emc just had one updating. Can you sync your branch ? I will reach out NCAR after you sync your branch. Thanks

@jiandewang , I've updated the branch. Thanks again for testing and the help.

**Comment by jiandewang** [comment]
now we got greenlight from NCAR. I will prepare a UWM PR for it to get merged to dev/emc

**Review by DeniseWorthen — APPROVED** [review]

**Review by jiandewang — APPROVED** [review]

**Comment by jiandewang** [comment]
combined with UWM (https://github.com/ufs-community/ufs-weather-model/pull/2205)

**Comment by FernandoAndrade-NOAA** [comment]
Testing for [#2205](https://github.com/ufs-community/ufs-weather-model/pull/2205) has completed successfully, please continue with merging this PR.

**Pull request merged**